stdlib/asm/collections/mmr.masm

```diff
@@ -104,3 +106,76 @@ export.get
      # stack: [leaf, ...]
   end
 end
+
+#! Load the MMR peak data based on its hash.
+#!
+#! Input: [HASH, mmr_ptr, ...]
+#! Output: [...]
+#!
+#! Where:
+#!  - HASH: is the MMR peak hash, the hash is expected to be padded to an even
+#!    length and to have a minimum size of 16 elements
+#!  - mmt_ptr: the memory location where the MMR data will be written to,
+#!    starting with the MMR forest (its total leaves count) followed by its peaks
+#!
+#! Cycles: 159 + 9 extra_peak_pair cycles
+#!    where `extra_peak` is the number of peak pairs in addition to the first
+#!    16, i.e. `round_up((num_of_peaks - 16) / 2)`
+export.unhash
+  # load the peaks to the advice_stack (0 cycles)
+  adv.keyval
+  # operand_stack => [HASH, mmr_ptr, ...]
+  # advice_stack => [peaks*, ...]
+
+  # load the number of leaves in the MMR to the advice_stack, the number of
+  # leaves is defined by the hash of the MMR peaks + 1 (3 cycles)
+  add.1 adv.keyval sub.1
+  # operand_stack => [HASH, mmr_ptr, ...]
+  # advice_stack => [forest, peaks*, ...]
```

Comment:
So `adv.keyvalnum` would remove the need for the `add.1` and `sub.1` and save two cycles?

Reply:
The main benefit is that we don't need to keep track of the number of elements under a separate key and thus avoid potential bugs when for some reason there is a divergence between the actual and stored number of elements.

Reply:
I would still prefer to add a different decorator for this as I think it places less of a burden on the user and reduces probability of bugs somewhat.

Reply:
Actually, for this specific PR this decorator is probably not useful as we get the number of peaks rather than the number of leaves.

Reply:
So, I guess we could go with:
```
advice_map = {
    HASH: [num_leaves, p1, p2, p3, p4, ...]
}
```
Not sure if this will complicate or simplify packing and insertions (seems like may simplify, but also I didn't really spend much time thinking about it).

Reply:
@bobbinth , to clarify, are you requesting to change this PR to implement the encoding above?

Reply:
I'm actually not sure because we need to look at how this will impact `pack` procedure. We should come up with an encoding which will work for both - but I don't know yet what it is.

Reply:
> So, I guess we could go with:
> 
> ```
> advice_map = {
>     HASH: [num_leaves, p1, p2, p3, p4, ...]
> }
> ```

This looks good to me.

Regarding `adv.keyvalnum`, I'm not sure if it should be implemented.  I think it should be the responsibility of the rust object to encode the data as `Felt`'s according to some defined encoding contract.  The masm code should be able to interpret it according to the contract.

What I think would be useful is constant support for the advice provider. I would like to be able to define constants in masm and then use them to specify data I want to be loaded from the advice provider map. 

```
const.ADV_MAP_KEY=1234
adv.keyval.ADV_MAP_KEY
```

Reply:
> It is packing to the operand stack, not the advice stack, which is probably what you're after.

There are two parts to this:
1. We need to compute the hash of all peaks.
2. We need to write the peaks back into the advice map and put them under the computed hash.

The second part could be a part of the `pack` procedure or it could be separate (i.e., a part of the insert `append` procedure). I guess we can keep it separate for now and revisit if we need to make any changes when we work on the `append` procedure.

In this case, yes, let's change the map to look like:
```
advice_map = {
    HASH: [num_leaves, p1, p2, p3, p4, ...]
}
```

Reply:
> Regarding `adv.keyvalnum`, I'm not sure if it should be implemented. I think it should be the responsibility of the rust object to encode the data as `Felt`'s according to some defined encoding contract. The masm code should be able to interpret it according to the contract.

Yes, let's hold off on this for now. I still think it would be useful - but not a priority any more.



> What I think would be useful is constant support for the advice provider. I would like to be able to define constants in masm and then use them to specify data I want to be loaded from the advice provider map.
> 
> ```
> const.ADV_MAP_KEY=1234
> adv.keyval.ADV_MAP_KEY
> ```

This could be interesting - but wouldn't keys be variable (rather than constant) most of the time? Are there any specific examples you have in mind where keys would be constant?

Reply:
> In this case, yes, let's change the map to look like:

updated

Reply:
> This could be interesting - but wouldn't keys be variable (rather than constant) most of the time? Are there any specific examples you have in mind where keys would be constant?

Any advice data that is known at compile time can be loaded into standardised map keys.  We have some scenarios in the kernel. For example, we know the global inputs, account data, block data, consumed note data ahead of time and load it in ordered fashion on the advice stack before execution.  Instead you could load it into standard keys in the advice map and call it from masm - something like this.

```rust
AdviceInputs {
    stack: [],
    map: {
       [0,0,0,0]: [GLOBAL_INPUTS],
       [1,0,0,0]: [ACCOUNT_DATA],
       [2,0,0,0]: [BLOCK_DATA],
    },
    store: MerkleStore::default(),
}
```

```
const.GLOBAL_INPUTS_KEY=0
const.ACCOUNT_DATA_KEY=1
const.BLOCK_DATA_KEY=2

proc.exec.process_global_inputs
      adv.keyval.GLOBAL_INPUTS_KEY
      ...
      adv_pipe adv_pipe
end

export.prepare_transaction
    # process global inputs
    exec.process_global_inputs

    # process block data
    exec.process_block_data

    ...
end

```

Reply:
> Any advice data that is known at compile time can be loaded into standardised map keys. We have some scenarios in the kernel. For example, we know the global inputs, account data, block data, consumed note data etc ahead of time and load it in ordered fashion on the advice stack before execution. Instead you could load it into standard keys in the advice map and call it from masm - something like this.

But aren't most of these bound to their own hashes. For example, account data is bound to account hash, block data is bound to block hash (and I think global inputs is a combination of block data, account, and note commitments). So, wouldn't it look more like this:
```Rust
AdviceInputs {
    stack: [],
    map: {
       acct_hash: [ACCOUNT_DATA],
       block_hash: [BLOCK_DATA],
       input_note_hash: [INPUT_NOTE_DATA],
    },
    store: MerkleStore::default(),
}
```
I do think there are probably cases where we do want to have some static key mapped to dynamic values, but I'm not sure came across this use case yet.

Reply:
> But aren't most of these bound to their own hashes. For example, account data is bound to account hash, block data is bound to block hash (and I think global inputs is a combination of block data, account, and note commitments). So, wouldn't it look more like this:
> 
> ```rust
> AdviceInputs {
>     stack: [],
>     map: {
>        acct_hash: [ACCOUNT_DATA],
>        block_hash: [BLOCK_DATA],
>        input_note_hash: [INPUT_NOTE_DATA],
>     },
>     store: MerkleStore::default(),
> }
> ```
> 
> I do think there are probably cases where we do want to have some static key mapped to dynamic values, but I'm not sure came across this use case yet.

I think it works fine as is for now however it forces the developer to load the key onto the stack before you can load data onto from the advice map.  I think we should be able to load data using well known keys without having to load the key onto the operand stack.  I don't think it's a big deal and probably not high priority but I do think it would be a useful feature.